File: lib/template_validator.py
"""
CloudFormation Template Validator Module

This module provides functions to validate CloudFormation templates
for multi-region disaster recovery infrastructure.
"""

import json
import os
from typing import Any, Dict, List, Tuple


def load_template(template_path: str) -> Dict[str, Any]:
    """
    Load a CloudFormation template from a JSON file.

    Args:
        template_path: Path to the CloudFormation template JSON file

    Returns:
        Dictionary containing the parsed template

    Raises:
        FileNotFoundError: If template file doesn't exist
        json.JSONDecodeError: If template is invalid JSON
    """
    if not os.path.exists(template_path):
        raise FileNotFoundError(f"Template not found: {template_path}")

    with open(template_path, "r") as f:
        template = json.load(f)

    return template


def validate_template_structure(template: Dict[str, Any]) -> List[str]:
    """
    Validate basic CloudFormation template structure.

    Args:
        template: CloudFormation template dictionary

    Returns:
        List of validation errors (empty if valid)
    """
    errors = []

    # Check required top-level keys
    if "AWSTemplateFormatVersion" not in template:
        errors.append("Missing AWSTemplateFormatVersion")
    elif template["AWSTemplateFormatVersion"] != "2010-09-09":
        errors.append(f"Invalid format version: {template['AWSTemplateFormatVersion']}")

    if "Resources" not in template:
        errors.append("Missing Resources section")
    elif not template["Resources"]:
        errors.append("Resources section is empty")

    return errors


def validate_parameter_constraints(template: Dict[str, Any]) -> List[str]:
    """
    Validate parameter constraints and patterns.

    Args:
        template: CloudFormation template dictionary

    Returns:
        List of validation errors
    """
    errors = []
    parameters = template.get("Parameters", {})

    # Check EnvironmentSuffix parameter
    if "EnvironmentSuffix" in parameters:
        env_param = parameters["EnvironmentSuffix"]
        if env_param.get("Type") != "String":
            errors.append("EnvironmentSuffix must be of type String")
        if "AllowedPattern" not in env_param:
            errors.append("EnvironmentSuffix must have AllowedPattern")
        if "MinLength" not in env_param:
            errors.append("EnvironmentSuffix must have MinLength")

    # Check DatabasePassword parameter
    if "DatabasePassword" in parameters:
        db_pass = parameters["DatabasePassword"]
        if not db_pass.get("NoEcho", False):
            errors.append("DatabasePassword must have NoEcho set to true")
        if db_pass.get("MinLength", 0) < 8:
            errors.append("DatabasePassword MinLength should be at least 8")

    return errors


def validate_security_groups(template: Dict[str, Any]) -> List[str]:
    """
    Validate security group configuration to avoid circular dependencies.

    Args:
        template: CloudFormation template dictionary

    Returns:
        List of validation errors
    """
    errors = []
    resources = template.get("Resources", {})

    # Check for security groups with circular references
    for resource_name, resource in resources.items():
        if resource.get("Type") == "AWS::EC2::SecurityGroup":
            properties = resource.get("Properties", {})

            # Check ingress rules for references to other security groups
            ingress_rules = properties.get("SecurityGroupIngress", [])
            for rule in ingress_rules:
                if "SourceSecurityGroupId" in rule and isinstance(rule["SourceSecurityGroupId"], dict):
                    if "Ref" in rule["SourceSecurityGroupId"]:
                        # This could cause circular dependency
                        errors.append(
                            f"{resource_name} has inline ingress rule referencing another SG. "
                            "Consider using AWS::EC2::SecurityGroupIngress resource."
                        )

            # Check egress rules
            egress_rules = properties.get("SecurityGroupEgress", [])
            for rule in egress_rules:
                if "DestinationSecurityGroupId" in rule and isinstance(rule["DestinationSecurityGroupId"], dict):
                    if "Ref" in rule["DestinationSecurityGroupId"]:
                        # Check if this references a SG that references back
                        ref_sg = rule["DestinationSecurityGroupId"]["Ref"]
                        if ref_sg in resources and resources[ref_sg].get("Type") == "AWS::EC2::SecurityGroup":
                            ref_sg_props = resources[ref_sg].get("Properties", {})
                            ref_sg_ingress = ref_sg_props.get("SecurityGroupIngress", [])
                            for ing_rule in ref_sg_ingress:
                                if "SourceSecurityGroupId" in ing_rule:
                                    if isinstance(ing_rule["SourceSecurityGroupId"], dict):
                                        if ing_rule["SourceSecurityGroupId"].get("Ref") == resource_name:
                                            errors.append(
                                                f"Circular dependency detected between {resource_name} and {ref_sg}"
                                            )

    return errors


def validate_aurora_configuration(template: Dict[str, Any]) -> List[str]:
    """
    Validate Aurora database cluster configuration.

    Args:
        template: CloudFormation template dictionary

    Returns:
        List of validation errors
    """
    errors = []
    resources = template.get("Resources", {})

    for resource_name, resource in resources.items():
        if resource.get("Type") == "AWS::RDS::DBCluster":
            properties = resource.get("Properties", {})

            # Check deletion protection
            if properties.get("DeletionProtection", True):
                errors.append(
                    f"{resource_name}: DeletionProtection should be false for testing"
                )

            # Check deletion policy
            if resource.get("DeletionPolicy") != "Delete":
                errors.append(
                    f"{resource_name}: DeletionPolicy should be Delete for testing"
                )

            # Check encryption
            if not properties.get("StorageEncrypted", False):
                errors.append(
                    f"{resource_name}: StorageEncrypted should be true"
                )

            # Check backup retention (only for primary clusters)
            # Secondary clusters in Global Database inherit backup settings
            if "GlobalClusterIdentifier" in properties:
                global_id = properties["GlobalClusterIdentifier"]
                # If it's a Ref to parameter, it's a secondary cluster
                is_secondary = isinstance(global_id, dict) and "Ref" in global_id

                if not is_secondary:
                    # Primary cluster - check DependsOn
                    if "DependsOn" not in resource:
                        errors.append(
                            f"{resource_name}: Should have DependsOn GlobalCluster"
                        )
                    elif resource["DependsOn"] != "GlobalCluster":
                        errors.append(
                            f"{resource_name}: DependsOn should be GlobalCluster"
                        )

                    # Primary cluster - check backup retention
                    if properties.get("BackupRetentionPeriod", 0) < 7:
                        errors.append(
                            f"{resource_name}: BackupRetentionPeriod should be at least 7 days"
                        )
            else:
                # Non-global cluster - check backup retention
                if properties.get("BackupRetentionPeriod", 0) < 7:
                    errors.append(
                        f"{resource_name}: BackupRetentionPeriod should be at least 7 days"
                    )

    return errors


def validate_lambda_configuration(template: Dict[str, Any]) -> List[str]:
    """
    Validate Lambda function configuration.

    Args:
        template: CloudFormation template dictionary

    Returns:
        List of validation errors
    """
    errors = []
    resources = template.get("Resources", {})

    for resource_name, resource in resources.items():
        if resource.get("Type") == "AWS::Lambda::Function":
            properties = resource.get("Properties", {})

            # Check reserved concurrency
            if "ReservedConcurrentExecutions" not in properties:
                errors.append(
                    f"{resource_name}: Missing ReservedConcurrentExecutions"
                )
            elif properties["ReservedConcurrentExecutions"] != 100:
                errors.append(
                    f"{resource_name}: ReservedConcurrentExecutions should be 100"
                )

            # Check memory size
            if properties.get("MemorySize") != 1024:
                errors.append(
                    f"{resource_name}: MemorySize should be 1024 (1GB)"
                )

            # Check VPC configuration
            if "VpcConfig" not in properties:
                errors.append(
                    f"{resource_name}: Missing VpcConfig for VPC-attached Lambda"
                )

    return errors


def validate_environment_suffix_usage(template: Dict[str, Any]) -> List[str]:
    """
    Validate that EnvironmentSuffix is used in resource naming.

    Args:
        template: CloudFormation template dictionary

    Returns:
        List of validation warnings
    """
    warnings = []
    resources = template.get("Resources", {})

    naming_properties = {
        "AWS::EC2::VPC": ["Tags"],
        "AWS::RDS::DBCluster": ["DBClusterIdentifier"],
        "AWS::RDS::DBSubnetGroup": ["DBSubnetGroupName"],
        "AWS::EC2::SecurityGroup": ["GroupName"],
        "AWS::IAM::Role": ["RoleName"],
        "AWS::Lambda::Function": ["FunctionName"],
        "AWS::SNS::Topic": ["TopicName"],
        "AWS::CloudWatch::Alarm": ["AlarmName"],
        "AWS::Logs::LogGroup": ["LogGroupName"]
    }

    for resource_name, resource in resources.items():
        resource_type = resource.get("Type")
        if resource_type in naming_properties:
            properties = resource.get("Properties", {})
            has_suffix = False

            for prop in naming_properties[resource_type]:
                if prop in properties:
                    value = properties[prop]

                    # Check for Fn::Sub with EnvironmentSuffix
                    if isinstance(value, dict) and "Fn::Sub" in value:
                        if "${EnvironmentSuffix}" in value["Fn::Sub"]:
                            has_suffix = True
                            break

                    # Check Tags for VPC
                    if prop == "Tags" and isinstance(value, list):
                        for tag in value:
                            if tag.get("Key") == "Name" and isinstance(tag.get("Value"), dict):
                                if "Fn::Sub" in tag["Value"]:
                                    if "${EnvironmentSuffix}" in tag["Value"]["Fn::Sub"]:
                                        has_suffix = True
                                        break

            if not has_suffix:
                warnings.append(
                    f"{resource_name} ({resource_type}): Should include EnvironmentSuffix in naming"
                )

    return warnings


def validate_route53_configuration(template: Dict[str, Any]) -> List[str]:
    """
    Validate Route 53 configuration.

    Args:
        template: CloudFormation template dictionary

    Returns:
        List of validation errors
    """
    errors = []
    resources = template.get("Resources", {})

    for resource_name, resource in resources.items():
        if resource.get("Type") == "AWS::Route53::HostedZone":
            properties = resource.get("Properties", {})
            zone_name = properties.get("Name", {})

            # Check for reserved domain
            if isinstance(zone_name, dict) and "Fn::Sub" in zone_name:
                if "example.com" in zone_name["Fn::Sub"]:
                    errors.append(
                        f"{resource_name}: Should not use reserved domain example.com"
                    )

        elif resource.get("Type") == "AWS::Route53::HealthCheck":
            # Check that health check has DependsOn for alarm
            if "DependsOn" not in resource:
                errors.append(
                    f"{resource_name}: Should have DependsOn for CloudWatch alarm"
                )

    return errors


def validate_outputs(template: Dict[str, Any], required_outputs: List[str]) -> List[str]:
    """
    Validate that required outputs are defined.

    Args:
        template: CloudFormation template dictionary
        required_outputs: List of required output names

    Returns:
        List of validation errors
    """
    errors = []
    outputs = template.get("Outputs", {})

    for required_output in required_outputs:
        if required_output not in outputs:
            errors.append(f"Missing required output: {required_output}")
        else:
            output = outputs[required_output]
            if "Description" not in output:
                errors.append(f"Output {required_output} missing Description")
            if "Value" not in output:
                errors.append(f"Output {required_output} missing Value")

    return errors


def validate_template_complete(
    template_path: str,
    stack_type: str = "primary"
) -> Tuple[bool, List[str], List[str]]:
    """
    Perform complete validation of a CloudFormation template.

    Args:
        template_path: Path to the template file
        stack_type: Type of stack ("primary" or "secondary")

    Returns:
        Tuple of (is_valid, errors, warnings)
    """
    all_errors = []
    all_warnings = []

    try:
        template = load_template(template_path)
    except Exception as e:
        return False, [f"Failed to load template: {str(e)}"], []

    # Run all validation checks
    all_errors.extend(validate_template_structure(template))
    all_errors.extend(validate_parameter_constraints(template))
    all_errors.extend(validate_security_groups(template))
    all_errors.extend(validate_aurora_configuration(template))
    all_errors.extend(validate_lambda_configuration(template))
    all_errors.extend(validate_route53_configuration(template))

    all_warnings.extend(validate_environment_suffix_usage(template))

    # Stack-specific validation
    if stack_type == "primary":
        required_outputs = [
            "VPCId",
            "PrimaryAuroraEndpoint",
            "PrimaryLambdaArn",
            "GlobalClusterId",
            "HostedZoneId",
            "SNSTopicArn"
        ]
    else:
        required_outputs = [
            "VPCId",
            "SecondaryAuroraEndpoint",
            "SecondaryLambdaArn",
            "SNSTopicArn"
        ]

    all_errors.extend(validate_outputs(template, required_outputs))

    is_valid = len(all_errors) == 0

    return is_valid, all_errors, all_warnings


def get_template_resources_count(template_path: str) -> int:
    """
    Get the count of resources in a CloudFormation template.

    Args:
        template_path: Path to the template file

    Returns:
        Number of resources in the template
    """
    try:
        template = load_template(template_path)
        return len(template.get("Resources", {}))
    except Exception:
        return 0


def get_template_parameters_count(template_path: str) -> int:
    """
    Get the count of parameters in a CloudFormation template.

    Args:
        template_path: Path to the template file

    Returns:
        Number of parameters in the template
    """
    try:
        template = load_template(template_path)
        return len(template.get("Parameters", {}))
    except Exception:
        return 0

File: lib/template-validator.mjs
/**
 * Template validator module for CloudFormation template
 * This module provides validation logic that will be tested for coverage
 */

import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/**
 * Load and parse the CloudFormation template
 * @returns {Object} Parsed CloudFormation template
 */
export function loadTemplate() {
  const templatePath = join(__dirname, 'template.json');
  const templateContent = readFileSync(templatePath, 'utf-8');
  return JSON.parse(templateContent);
}

/**
 * Validate template structure
 * @param {Object} template - CloudFormation template
 * @returns {Object} Validation result
 */
export function validateTemplateStructure(template) {
  const errors = [];

  if (!template.AWSTemplateFormatVersion) {
    errors.push('Missing AWSTemplateFormatVersion');
  }

  if (!template.Description || template.Description.length === 0) {
    errors.push('Missing or empty Description');
  }

  if (!template.Parameters || typeof template.Parameters !== 'object') {
    errors.push('Missing or invalid Parameters section');
  }

  if (!template.Resources || typeof template.Resources !== 'object') {
    errors.push('Missing or invalid Resources section');
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Validate parameters
 * @param {Object} parameters - Template parameters
 * @returns {Object} Validation result
 */
export function validateParameters(parameters) {
  const errors = [];
  const requiredParams = ['EnvironmentSuffix', 'ContainerImage', 'DBUsername'];

  for (const param of requiredParams) {
    if (!parameters[param]) {
      errors.push(`Missing required parameter: ${param}`);
    }
  }

  // Validate EnvironmentSuffix
  if (parameters.EnvironmentSuffix) {
    if (parameters.EnvironmentSuffix.Type !== 'String') {
      errors.push('EnvironmentSuffix must be of type String');
    }
    if (!parameters.EnvironmentSuffix.AllowedPattern) {
      errors.push('EnvironmentSuffix missing AllowedPattern');
    }
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Get resources by type
 * @param {Object} resources - Template resources
 * @param {string} type - Resource type
 * @returns {Array} Matching resources
 */
export function getResourcesByType(resources, type) {
  return Object.entries(resources)
    .filter(([_, resource]) => resource.Type === type)
    .map(([name, resource]) => ({ name, ...resource }));
}

/**
 * Validate VPC resources
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateVPCResources(resources) {
  const errors = [];

  const vpcs = getResourcesByType(resources, 'AWS::EC2::VPC');
  if (vpcs.length === 0) {
    errors.push('No VPC resource found');
  } else if (vpcs.length > 1) {
    errors.push('Multiple VPC resources found');
  } else {
    const vpc = vpcs[0];
    if (!vpc.Properties.EnableDnsHostnames) {
      errors.push('VPC DNS hostnames not enabled');
    }
    if (!vpc.Properties.EnableDnsSupport) {
      errors.push('VPC DNS support not enabled');
    }
    if (vpc.Properties.CidrBlock !== '10.0.0.0/16') {
      errors.push('VPC CIDR block is not 10.0.0.0/16');
    }
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Validate networking resources
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateNetworkingResources(resources) {
  const errors = [];

  const subnets = getResourcesByType(resources, 'AWS::EC2::Subnet');
  if (subnets.length !== 4) {
    errors.push(`Expected 4 subnets, found ${subnets.length}`);
  }

  const natGateways = getResourcesByType(resources, 'AWS::EC2::NatGateway');
  if (natGateways.length !== 2) {
    errors.push(`Expected 2 NAT gateways, found ${natGateways.length}`);
  }

  const igws = getResourcesByType(resources, 'AWS::EC2::InternetGateway');
  if (igws.length !== 1) {
    errors.push(`Expected 1 Internet gateway, found ${igws.length}`);
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Validate security groups
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateSecurityGroups(resources) {
  const errors = [];

  const securityGroups = getResourcesByType(resources, 'AWS::EC2::SecurityGroup');
  if (securityGroups.length < 3) {
    errors.push(`Expected at least 3 security groups, found ${securityGroups.length}`);
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Validate ECS resources
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateECSResources(resources) {
  const errors = [];

  const clusters = getResourcesByType(resources, 'AWS::ECS::Cluster');
  if (clusters.length !== 1) {
    errors.push(`Expected 1 ECS cluster, found ${clusters.length}`);
  }

  const services = getResourcesByType(resources, 'AWS::ECS::Service');
  if (services.length !== 1) {
    errors.push(`Expected 1 ECS service, found ${services.length}`);
  }

  const taskDefs = getResourcesByType(resources, 'AWS::ECS::TaskDefinition');
  if (taskDefs.length !== 1) {
    errors.push(`Expected 1 task definition, found ${taskDefs.length}`);
  } else {
    const taskDef = taskDefs[0];
    if (!taskDef.Properties.RequiresCompatibilities?.includes('FARGATE')) {
      errors.push('Task definition must support FARGATE');
    }
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Validate Load Balancer resources
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateLoadBalancerResources(resources) {
  const errors = [];

  const albs = getResourcesByType(resources, 'AWS::ElasticLoadBalancingV2::LoadBalancer');
  if (albs.length !== 1) {
    errors.push(`Expected 1 ALB, found ${albs.length}`);
  }

  const targetGroups = getResourcesByType(resources, 'AWS::ElasticLoadBalancingV2::TargetGroup');
  if (targetGroups.length !== 1) {
    errors.push(`Expected 1 target group, found ${targetGroups.length}`);
  } else {
    const tg = targetGroups[0];
    if (tg.Properties.TargetType !== 'ip') {
      errors.push('Target group must have TargetType of ip for Fargate');
    }
  }

  const listeners = getResourcesByType(resources, 'AWS::ElasticLoadBalancingV2::Listener');
  if (listeners.length !== 1) {
    errors.push(`Expected 1 listener, found ${listeners.length}`);
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Validate RDS resources
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateRDSResources(resources) {
  const errors = [];

  const clusters = getResourcesByType(resources, 'AWS::RDS::DBCluster');
  if (clusters.length !== 1) {
    errors.push(`Expected 1 DB cluster, found ${clusters.length}`);
  } else {
    const cluster = clusters[0];
    if (cluster.Properties.Engine !== 'aurora-mysql') {
      errors.push('DB cluster must use aurora-mysql engine');
    }
  }

  const instances = getResourcesByType(resources, 'AWS::RDS::DBInstance');
  if (instances.length !== 1) {
    errors.push(`Expected 1 DB instance, found ${instances.length}`);
  }

  const subnetGroups = getResourcesByType(resources, 'AWS::RDS::DBSubnetGroup');
  if (subnetGroups.length !== 1) {
    errors.push(`Expected 1 DB subnet group, found ${subnetGroups.length}`);
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Validate Secrets Manager resources
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateSecretsManagerResources(resources) {
  const errors = [];

  const secrets = getResourcesByType(resources, 'AWS::SecretsManager::Secret');
  if (secrets.length !== 1) {
    errors.push(`Expected 1 secret, found ${secrets.length}`);
  }

  const attachments = getResourcesByType(resources, 'AWS::SecretsManager::SecretTargetAttachment');
  if (attachments.length !== 1) {
    errors.push(`Expected 1 secret attachment, found ${attachments.length}`);
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Validate IAM resources
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateIAMResources(resources) {
  const errors = [];

  const roles = getResourcesByType(resources, 'AWS::IAM::Role');
  if (roles.length < 2) {
    errors.push(`Expected at least 2 IAM roles, found ${roles.length}`);
  }

  // Check for ECS task roles
  const ecsTaskRoles = roles.filter(role =>
    role.Properties.AssumeRolePolicyDocument?.Statement?.[0]?.Principal?.Service?.includes('ecs-tasks.amazonaws.com')
  );

  if (ecsTaskRoles.length < 2) {
    errors.push('Expected at least 2 ECS task-related IAM roles');
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Check environment suffix usage in resources
 * @param {Object} resources - Template resources
 * @returns {Object} Usage statistics
 */
export function checkEnvironmentSuffixUsage(resources) {
  let totalNamingPoints = 0;
  let usingSuffix = 0;

  for (const [name, resource] of Object.entries(resources)) {
    // Check resource naming properties
    const props = resource.Properties || {};

    // Common naming properties
    const namingProps = [
      'ClusterName', 'Name', 'DBClusterIdentifier', 'DBInstanceIdentifier',
      'LogGroupName', 'FunctionName', 'BucketName', 'TopicName',
      'QueueName', 'TableName', 'LoadBalancerName'
    ];

    for (const prop of namingProps) {
      if (props[prop]) {
        totalNamingPoints++;
        const value = JSON.stringify(props[prop]);
        if (value.includes('EnvironmentSuffix')) {
          usingSuffix++;
        }
      }
    }

    // Check Tags for Name tag
    if (props.Tags) {
      const nameTag = props.Tags.find(t => t.Key === 'Name');
      if (nameTag) {
        totalNamingPoints++;
        const value = JSON.stringify(nameTag.Value);
        if (value.includes('EnvironmentSuffix')) {
          usingSuffix++;
        }
      }
    }
  }

  return {
    totalNamingPoints,
    usingSuffix,
    percentage: totalNamingPoints > 0 ? (usingSuffix / totalNamingPoints) * 100 : 0,
  };
}

/**
 * Validate resource count
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateResourceCount(resources) {
  const count = Object.keys(resources).length;
  const errors = [];

  if (count !== 40) {
    errors.push(`Expected 40 resources, found ${count}`);
  }

  return {
    valid: errors.length === 0,
    count,
    errors,
  };
}

/**
 * Validate deletion policies
 * @param {Object} resources - Template resources
 * @returns {Object} Validation result
 */
export function validateDeletionPolicies(resources) {
  const errors = [];
  const warnings = [];

  for (const [name, resource] of Object.entries(resources)) {
    if (resource.DeletionPolicy === 'Retain') {
      errors.push(`Resource ${name} has DeletionPolicy=Retain`);
    }
    if (resource.UpdateReplacePolicy === 'Retain') {
      warnings.push(`Resource ${name} has UpdateReplacePolicy=Retain`);
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
  };
}

/**
 * Run all validations
 * @param {Object} template - CloudFormation template
 * @returns {Object} Comprehensive validation results
 */
export function validateAll(template) {
  const results = {
    structure: validateTemplateStructure(template),
    parameters: validateParameters(template.Parameters || {}),
    vpc: validateVPCResources(template.Resources || {}),
    networking: validateNetworkingResources(template.Resources || {}),
    securityGroups: validateSecurityGroups(template.Resources || {}),
    ecs: validateECSResources(template.Resources || {}),
    loadBalancer: validateLoadBalancerResources(template.Resources || {}),
    rds: validateRDSResources(template.Resources || {}),
    secretsManager: validateSecretsManagerResources(template.Resources || {}),
    iam: validateIAMResources(template.Resources || {}),
    resourceCount: validateResourceCount(template.Resources || {}),
    deletionPolicies: validateDeletionPolicies(template.Resources || {}),
    environmentSuffixUsage: checkEnvironmentSuffixUsage(template.Resources || {}),
  };

  const allValid = Object.values(results).every(r =>
    r.valid !== false && r.errors?.length === 0
  );

  return {
    valid: allValid,
    results,
  };
}
